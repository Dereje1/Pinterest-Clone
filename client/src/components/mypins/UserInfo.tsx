import React, { useState } from 'react';
import Avatar from '@mui/material/Avatar';
import EditIcon from '@mui/icons-material/Edit';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import TextField from '@mui/material/TextField';
import DoneIcon from '@mui/icons-material/Done';
import CancelIcon from '@mui/icons-material/Cancel';
import { getProviderIcons, UserJoinedDate } from '../common/common';
import {
  providerIconsType,
} from '../../interfaces';

const providerIcons = getProviderIcons({ fontSize: 45 });

interface UserInfoProps {
  user:{
    service: string
    displayName: string | null
    username: string | null
    createdAt: string
  }
  showNameChangeForm: ()=> void
  nameChangeFormIsShowing: boolean
  submitNameChange: (newName: string)=>void
}

interface NameChangeFormProps {
    submitNameChange: (newName: string)=>void
    oldDisplayName: string | null,
  }

export function NameChangeForm({ submitNameChange, oldDisplayName }: NameChangeFormProps) {
  const [nameValue, setNameValue] = useState(oldDisplayName || '');

  const handleNameChange = (e: React.SyntheticEvent, isCancelled = false) => {
    if (!isCancelled) {
      submitNameChange(nameValue.trim());
    } else {
      submitNameChange('');
    }
  };

  return (
    <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}>
      <TextField
        variant="standard"
        value={nameValue}
        onChange={({ target: { value } }) => setNameValue(value)}
        error={nameValue.trim().length > 15}
        sx={{ width: 150, mb: 1 }}
        InputProps={{
          inputProps: {
            style: { textAlign: 'center' },
          },
        }}
      />
      <div style={{
        display: 'flex', justifyContent: 'space-around', width: 150,
      }}
      >
        <IconButton onClick={handleNameChange}>
          <DoneIcon
            color="success"
          />
        </IconButton>
        <IconButton onClick={(_) => handleNameChange(_, true)}>
          <CancelIcon
            color="error"
          />
        </IconButton>
      </div>
    </div>
  );
}

function UserInfo({
  user,
  nameChangeFormIsShowing,
  showNameChangeForm,
  submitNameChange,
}: UserInfoProps) {
  const { service, displayName, username } = user;
  return (
    <div style={{
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      justifyContent: 'center',
    }}
    >
      <Avatar sx={{
        mb: 2,
        width: 52,
        height: 52,
        bgcolor: providerIcons[service as keyof providerIconsType].color,
      }}
      >
        {providerIcons[service as keyof providerIconsType].icon}
      </Avatar>
      <div style={{ display: 'flex', alignItems: 'center', height: 75 }}>
        {nameChangeFormIsShowing ? (
          <NameChangeForm
            submitNameChange={submitNameChange}
            oldDisplayName={displayName}
          />
        ) : (
          <>
            <Typography variant="h5" sx={{ paddingLeft: 8 }}>
              {displayName}
            </Typography>
            <IconButton onClick={showNameChangeForm} sx={{ ml: 2 }}>
              <EditIcon fontSize="small" />
            </IconButton>
          </>
        )}
      </div>
      <Typography style={{ marginLeft: 15, marginTop: 4 }}>
        {service === 'google' ? username : `@${username}`}
      </Typography>
      <UserJoinedDate dateJoined={user.createdAt} />
    </div>
  );
}

export default UserInfo;
