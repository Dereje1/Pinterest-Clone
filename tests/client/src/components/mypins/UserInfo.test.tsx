import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import UserInfo, { NameChangeForm } from '../../../../../client/src/components/mypins/UserInfo';

describe('The UserInfo component', () => {
  let props;
  beforeEach(() => {
    props = {
      user: {
        service: 'twitter',
        displayName: 'tester displayName',
        username: 'tester username',
      },
      showNameChangeForm: jest.fn(),
      nameChangeFormIsShowing: false,
      submitNameChange: jest.fn(),
    };
  });

  afterEach(() => {
    props = null;
  });

  test('will render with display name', () => {
    const wrapper = shallow(<UserInfo {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    const userName = wrapper.find('ForwardRef(Typography)').at(1);
    expect(userName.text()).toBe('@tester username');
  });

  test('will render with change name form', () => {
    const updatedProps = { ...props, nameChangeFormIsShowing: true };
    const wrapper = shallow(<UserInfo {...updatedProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render with username with no @ symbol', () => {
    const updatedProps = {
      ...props,
      user: {
        ...props.user,
        service: 'google',
      },
    };
    const wrapper = shallow(<UserInfo {...updatedProps} />);
    const userName = wrapper.find('ForwardRef(Typography)').at(1);
    expect(userName.text()).toBe('tester username');
  });
});

describe('The NameChangeForm component', () => {
  let props;
  beforeEach(() => {
    props = {
      submitNameChange: jest.fn(),
      oldDisplayName: 'tester displayName',
    };
  });

  afterEach(() => {
    props = null;
  });

  test('will control the value of the name change', () => {
    const wrapper = shallow(<NameChangeForm {...props} />);
    let textField = wrapper.find('ForwardRef(TextField)');
    expect(textField.props().value).toBe('tester displayName');
    textField.props().onChange({ target: { value: 'new value' } });
    textField = wrapper.find('ForwardRef(TextField)');
    expect(textField.props().value).toBe('new value');
  });

  test('will set an empty string if original display name is null', () => {
    const updatedProps = { ...props, oldDisplayName: null };
    const wrapper = shallow(<NameChangeForm {...updatedProps} />);
    const textField = wrapper.find('ForwardRef(TextField)');
    expect(textField.props().value).toBe('');
  });

  test('will handle a submission of the changed name', () => {
    const wrapper = shallow(<NameChangeForm {...props} />);
    const textField = wrapper.find('ForwardRef(TextField)');
    expect(textField.props().value).toBe('tester displayName');
    textField.props().onChange({ target: { value: 'new value' } });
    const doneIcon = wrapper.find('ForwardRef(IconButton)').at(0);
    doneIcon.props().onClick();
    expect(props.submitNameChange).toHaveBeenCalledWith('new value');
  });

  test('will handle a cancellation for the changed name', () => {
    const wrapper = shallow(<NameChangeForm {...props} />);
    const textField = wrapper.find('ForwardRef(TextField)');
    expect(textField.props().value).toBe('tester displayName');
    textField.props().onChange({ target: { value: 'new value' } });
    const cancelIcon = wrapper.find('ForwardRef(IconButton)').at(1);
    cancelIcon.props().onClick('', true);
    expect(props.submitNameChange).toHaveBeenCalledWith('');
  });
});
