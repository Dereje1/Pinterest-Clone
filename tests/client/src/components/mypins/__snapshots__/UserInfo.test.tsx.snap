// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The UserInfo component will render with change name form 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <ForwardRef(Avatar)
    sx={
      {
        "bgcolor": "#1DA1F2",
        "height": 52,
        "mb": 2,
        "width": 52,
      }
    }
  >
    <Memo(ForwardRef(TwitterIcon))
      style={
        {
          "fontSize": 45,
        }
      }
    />
  </ForwardRef(Avatar)>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "height": 75,
      }
    }
  >
    <NameChangeForm
      oldDisplayName="tester displayName"
      submitNameChange={[MockFunction]}
    />
  </div>
  <ForwardRef(Typography)
    style={
      {
        "marginLeft": 15,
        "marginTop": 4,
      }
    }
  >
    @tester username
  </ForwardRef(Typography)>
</div>
`;

exports[`The UserInfo component will render with display name 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <ForwardRef(Avatar)
    sx={
      {
        "bgcolor": "#1DA1F2",
        "height": 52,
        "mb": 2,
        "width": 52,
      }
    }
  >
    <Memo(ForwardRef(TwitterIcon))
      style={
        {
          "fontSize": 45,
        }
      }
    />
  </ForwardRef(Avatar)>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "height": 75,
      }
    }
  >
    <ForwardRef(Typography)
      sx={
        {
          "paddingLeft": 8,
        }
      }
      variant="h5"
    >
      tester displayName
    </ForwardRef(Typography)>
    <ForwardRef(IconButton)
      onClick={[MockFunction]}
      sx={
        {
          "ml": 2,
        }
      }
    >
      <Memo(ForwardRef(EditIcon))
        fontSize="small"
      />
    </ForwardRef(IconButton)>
  </div>
  <ForwardRef(Typography)
    style={
      {
        "marginLeft": 15,
        "marginTop": 4,
      }
    }
  >
    @tester username
  </ForwardRef(Typography)>
</div>
`;
